File: T-Bank/Algorithms-and-Data-Structures-2025/11sprint-Graph2/contest/C-Печатная_схема/solution.cpp
#include <iostream>
#include <vector>
#include <tuple>


class UnionFind {
private:
    int n;
    std::vector<int> parent;
    std::vector<int> rank;
    std::vector<int> minValue;
    std::vector<int> maxValue;
    int components;

public:
    UnionFind(int n_) : n(n_) {
        parent.resize(n, 0);
        rank.resize(n, 1);
        minValue.resize(n, 0);
        maxValue.resize(n, 0);
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
            minValue[i] = i;
            maxValue[i] = i;
        }
        components = n;
    }
    
    int find(int x) {
        while (x != parent[x]) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }
    
    void union_(int v, int u) {
        int vParent = find(v), uParent = find(u);
        if (vParent == uParent) return;
        --components;
        if (rank[vParent] >= rank[uParent]) {
            parent[uParent] = vParent;
            rank[vParent] += rank[uParent];
            minValue[vParent] = std::min(minValue[vParent], minValue[uParent]);
            maxValue[vParent] = std::max(maxValue[vParent], maxValue[uParent]);
        } else {
            parent[vParent] = uParent;
            rank[uParent] += rank[vParent];
            minValue[uParent] = std::min(minValue[uParent], minValue[vParent]);
            maxValue[uParent] = std::max(maxValue[uParent], maxValue[vParent]);
        }
    }
};



void solution() {
    /*
    Time Complexity O((NM)log(NM))
    Memory Complexity O(NM)
    */
    int n, m;
    std::scanf("%d %d", &n, &m);
    UnionFind uf(n * m + 2);
    std::vector<std::vector<int>> grid (n + 1, std::vector<int>(m + 1, 0));
    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= m; ++j) std::scanf("%d", &grid[i][j]);
    }
    for (int r = 1; r <= n; ++r) {
        for (int c = 1; c <= m; ++c) {
            if (grid[r][c] == 1) {
                uf.union_((r - 1) * m + c, (r - 1) * m + c + m);
            } else if (grid[r][c] == 2) {
                uf.union_((r - 1) * m + c, (r - 1) * m + c + 1);
            } else if (grid[r][c] == 3) {
                uf.union_((r - 1) * m + c, (r - 1) * m + c + m);
                uf.union_((r - 1) * m + c, (r - 1) * m + c + 1);
            }
        }
    }
    int score = 0;
    std::vector<std::tuple<int, int, int>> moves;
    for (int r = 1; r < n; ++r) {
        for (int c = 1; c <= m; ++c) {
            if (uf.find((r - 1) * m + c) != uf.find((r - 1) * m + c + m)) {
                ++score;
                uf.union_((r - 1) * m + c, (r - 1) * m + c + m);
                moves.push_back({r, c, 1});
            }
        }
    }
    for (int r = 1; r <= n; ++r) {
        for (int c = 1; c < m; ++c) {
            if (uf.find((r - 1) * m + c) != uf.find((r - 1) * m + c + 1)) {
                score += 2;
                moves.push_back({r, c, 2});
                uf.union_((r - 1) * m + c, (r - 1) * m + c + 1);
            }
        }
    }
    std::printf("%d %d\n", moves.size(), score);
    for (std::tuple<int, int, int> &move : moves) {
        std::printf("%d %d %d\n", std::get<0>(move), std::get<1>(move), std::get<2>(move));
    }
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/11sprint-Graph2/contest/C-Печатная_схема/solution.py
import sys


class UnionFind:
    def __init__(self, n: int) -> None:
        self.parent: list[int] = list(range(n))
        self.rank: list[int] = [1] * n
        self.min_value: list[int] = list(range(n))
        self.max_value: list[int] = list(range(n))
    
    def find(self, x: int) -> int:
        while x != self.parent[x]:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x
    
    def union_(self, u: int, v: int) -> None:
        u_parent: int = self.find(u)
        v_parent: int = self.find(v)
        if u_parent == v_parent: return
        if self.rank[u_parent] >= self.rank[v_parent]:
            self.parent[v_parent] = u_parent
            self.rank[u_parent] += self.rank[v_parent]
            self.min_value[u_parent] = min(self.min_value[u_parent], self.min_value[v_parent])
            self.max_value[u_parent] = max(self.max_value[u_parent], self.max_value[v_parent])
        else:
            self.parent[u_parent] = v_parent
            self.rank[v_parent] += self.rank[u_parent]
            self.min_value[v_parent] = min(self.min_value[v_parent], self.min_value[u_parent])
            self.max_value[v_parent] = max(self.max_value[v_parent], self.max_value[u_parent])
    


def solution() -> None:
    """
    Time Complexity O((NM)log(NM))
    Memory Complexity O(NM)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    uf: UnionFind = UnionFind(n * m + 2)
    grid: list[list[int]] = [[]]
    for _ in range(n):
        row: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
        grid.append([0] + row)
    for r in range(1, n + 1):
        for c in range(1, m + 1):
            if grid[r][c] == 1:
                uf.union_((r - 1) * m + c, (r - 1) * m + c + m)
            elif grid[r][c] == 2:
                uf.union_((r - 1) * m + c, (r - 1) * m + c + 1)
            elif grid[r][c] == 3:
                uf.union_((r - 1) * m + c, (r - 1) * m + c + m)
                uf.union_((r - 1) * m + c, (r - 1) * m + c + 1)
    score: int = 0
    moves: list[tuple[int, int, int]] = []
    for r in range(1, n):
        for c in range(1, m + 1):
            if uf.find((r - 1) * m + c) != uf.find((r - 1) * m + c + m):
                score += 1
                moves.append((r, c, 1))
                uf.union_((r - 1) * m + c, (r - 1) * m + c + m)
    for r in range(1, n + 1):
        for c in range(1, m):
            if uf.find((r - 1) * m + c) != uf.find((r - 1) * m + c + 1):
                score += 2
                moves.append((r, c, 2))
                uf.union_((r - 1) * m + c, (r - 1) * m + c + 1)
    sys.stdout.write('{} {}\n'.format(len(moves), score))
    if not score: return
    sys.stdout.write('{}'.format('\n'.join((' '.join(map(str, move))) for move in moves)))

if __name__ == '__main__':
    solution()